src/jit/core_templates.expr

```diff
@@ -0,0 +1,229 @@
+(template: le_i (flagval (le $1 $2)))
+(template: lt_i (flagval (lt $1 $2)))
+(template: eqaddr (flagval (eq $1 $2)))
+
+
+(template: sp_getarg_o (load (^parg $1) ptr_sz))
+(template: sp_getarg_s (load (^parg $1) ptr_sz))
+(template: sp_getarg_i (load (^parg $1) int_sz))
+(template: sp_getarg_n (load (^parg $1) int_sz))
+
+
+(template: sp_getspeshslot (load (^spesh_slot $1) ptr_sz))
+
+(template: null_s (const 0 ptr_sz))
+
+(template: null (^vmnull))
+
+(template: getwhat (^getf (^stable $1) MVMSTable WHAT))
+
+
+(template: hllboxtype_i
+    (^getf (^hllconfig) MVMHLLConfig int_box_type))
+
+(template: hllboxtype_s
+    (^getf (^hllconfig) MVMHLLConfig str_box_type))
+
+(template: hllboxtype_n
+    (^getf (^hllconfig) MVMHLLConfig num_box_type))
+
+
+(template: curcode
+    (^getf (^frame) MVMFrame code_ref))
+
+
+# Relatively harmless branches, I think
+
+(template: unless_i
+    (when
+        (zr $0)
+        (branch $1)))
+
+(template: if_i
+    (when
+        (nz $0)
+        (branch $1)))
+
+(template: ifnonnull
+    (when (all
+             (nz $0) (ne $0 (^vmnull)))
+        (branch $1)))
+
+(template: goto (branch $0))
+
+
+# since compilation of invocation is special-cased in the regular jit
+# we can't just toss this in here without also handling all of arg_*
+# and invoke_*.
+#(template: prepargs (^setf (^getf (tc) MVMThreadContext cur_frame) MVMFrame cur_args_callsite
+#                    (^cu_callsite $0)))
+
+(template: sp_p6oget_i (load (add (^p6obody $1) $2) int_sz))
+(template: sp_p6oget_n (load (add (^p6obody $1) $2) (&sizeof MVMnum64)))
+(template: sp_p6oget_s (load (add (^p6obody $1) $2) ptr_sz))
+(template: sp_p6oget_o
+           (let: (($val (load (add (^p6obody $1) $2) ptr_sz)))
+                (if (nz $val) $val (^vmnull))))
+
+(template: getwho
+   (let: (($who (^getf (^stable $1) MVMSTable WHO)))
+     (if (nz $who) $who (^vmnull))))
+
+
+(template: sp_p6obind_i (store (add (^p6obody $0) $1) $2 int_sz))
+(template: sp_p6obind_n (store (add (^p6obody $0) $1) $2 int_sz))
+
+(template: sp_p6obind_o
+    (dov (^write_barrier $0 $2)
+         (store (add (^p6obody $0) $1) $2 ptr_sz)))
+
+(template: sp_p6obind_s
+    (dov (^write_barrier $0 $2)
+         (store (add (^p6obody $0) $1) $2 ptr_sz)))
+
+(template: sp_p6ogetvt_o
+   (let: (($addr (add (^p6obody $1) $2))
+          ($val (load $addr ptr_sz)))
+       (if (nz $val)
+            $val
+            (let: (($type (load (^spesh_slot $3) ptr_sz)))
+                 (^write_barrier $1 $type)
+                 (store $addr $type ptr_sz)
+                 (copy $type)))))
+
+(template: say (callv (^func &MVM_string_say)
+                      (arglist (carg (tc) ptr)
+                               (carg $0 ptr))))
+
+(template: decont!
+    (ifv
+         (any (zr $1)
+              (^is_type_obj $1)
+              (zr (^getf (^stable $1) MVMSTable container_spec)))
+         (store $0 $1 ptr_sz)
+         (callv (^stable_cont_func $1 fetch)
```

Comment:
I guess this'd be a bit more efficient with a let that dereferenced the container_spec off the STable once?

Reply:
Yes, but we'd have to guard against `$1` being non-null, and in general complicate the logic quite considerably... It'd be something like:

    (ifv (all (nz $1) (^is_concrete_obj $1)) 
          (let (($cont_spec (^getf (^stable $1) MVMSTable container_spec)))
                 (ifv (nz $const_spec)
                       (callv (^getf $const_spec MVMContainerSpec) 
                                 ...)
                       (store $0 $1 ptr_sz))
        (store $0 $1 ptr_sz))

And considering that due to tiling, the loads in `(^getf)` may be done doubly anyway (I'm planning an optimization to prevent that), I'm not sure we're actually doing better, and more importantly, the (future) optimizer should be able to resolve these things automatically.

But thanks for the comment, well found.